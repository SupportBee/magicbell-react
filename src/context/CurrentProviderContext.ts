import { createContext } from 'react';

export const NO_PROVIDER = 'MAGICBELL_NO_PROVIDER';

const CurrentProviderContext = createContext<String>(NO_PROVIDER);
export default CurrentProviderContext;
