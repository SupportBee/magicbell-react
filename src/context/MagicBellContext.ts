import { createContext, useContext } from 'react';

export interface IMagicBellContext {
  images?: Partial<{
    emptyInboxUrl: string;
  }>;
}

const MagicBellContext = createContext<IMagicBellContext>({} as IMagicBellContext);
export default MagicBellContext;

const useMagicBellContext = () => useContext(MagicBellContext);
export { useMagicBellContext };
