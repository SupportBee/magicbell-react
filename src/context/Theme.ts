export interface IElementTheme {
  backgroundColor: string;
  backgroundOpacity: number;
  borderRadius: string;
  fontFamily: string;
  fontSize: string;
  textAlign: 'left' | 'right' | 'center' | 'justify' | 'initial' | 'inherit';
  textColor: string;
  textTransform: 'none' | 'capitalize' | 'uppercase' | 'lowercase' | 'initial' | 'inherit';
}

export interface IMagicBellTheme {
  icon: {
    borderColor: string;
    width: string;
  };
  unseenBadge: IElementTheme;
  header: IElementTheme;
  container: IElementTheme;
  footer: IElementTheme;
  notification: {
    default: IElementTheme;
    unread: IElementTheme;
    unseen: IElementTheme;
  };
}

const blue = '#3498F4';
const red = '#DF4759';
const black = '#3A424D';

// prettier-ignore
const fontFamily = '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen-Sans, Ubuntu, Cantarell, "Helvetica Neue", sans-serif';
const fontSize = '14px';
const textAlign = 'left';
const textTransform = 'none';

export const defaultTheme = {
  icon: {
    borderColor: blue,
    width: '24px',
  },
  header: {
    backgroundColor: blue,
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize,
    textAlign,
    textColor: 'white',
    textTransform,
  },
  footer: {
    backgroundColor: blue,
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize: fontSize,
    textAlign,
    textColor: 'white',
    textTransform,
  },
  unseenBadge: {
    backgroundColor: red,
    backgroundOpacity: 1,
    borderRadius: '2px',
    fontFamily,
    fontSize: '10px',
    textAlign: 'center',
    textColor: 'white',
    textTransform,
  },
  container: {
    backgroundColor: '#FFFFFF',
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize: '14px',
    textAlign,
    textColor: black,
    textTransform,
  },
  notification: {
    default: {
      backgroundColor: blue,
      backgroundOpacity: 0.1,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
    unread: {
      backgroundColor: blue,
      backgroundOpacity: 0,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
    unseen: {
      backgroundColor: blue,
      backgroundOpacity: 0.05,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
  },
} as IMagicBellTheme;
