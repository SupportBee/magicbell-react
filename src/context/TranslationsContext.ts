import { createContext, useContext } from 'react';

const TranslationsContext = createContext<any>({});
export default TranslationsContext;

const TranslationsProvider = TranslationsContext.Provider;
const useTranslations = () => useContext(TranslationsContext);
export { TranslationsProvider, useTranslations };
