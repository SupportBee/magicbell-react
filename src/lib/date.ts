import dayjs from 'dayjs';
import localizedFormat from 'dayjs/plugin/localizedFormat';
import relativeTime from 'dayjs/plugin/relativeTime';
import updateLocale from 'dayjs/plugin/updateLocale';

dayjs.extend(localizedFormat);
dayjs.extend(relativeTime);
dayjs.extend(updateLocale);

dayjs.updateLocale('en', {
  relativeTime: {
    future: 'in %s',
    past: '%s ago',
    s: '%ds',
    m: '1m',
    mm: '%dm',
    h: '1h',
    hh: '%dh',
    d: '1d',
    dd: '%dd',
    M: '1mo',
    MM: '%dmo',
    y: '1y',
    yy: '%dy',
  },
});
