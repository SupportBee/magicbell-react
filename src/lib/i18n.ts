import dayjs from 'dayjs';
import 'dayjs/locale/es';
import 'dayjs/locale/pt-br';
import defaultsDeep from 'lodash/defaultsDeep';
import es from './translations/es';
import pt_br from './translations/pt-br';

export type CustomLocale = {
  name: string;
  translations: any;
};

function setDayjsLocale(localeName: string) {
  // All dayjs files are in lowercase and use hypens instead of underscores.
  const standardLocaleName = localeName.toLowerCase().replace('_', '-');
  dayjs.locale(standardLocaleName);
}

/**
 * Function to build localization data for the `TranslationsProvider`.
 *
 * @param locale
 */
export function useLocale(locale: string | CustomLocale) {
  const localeName = typeof locale === 'string' ? locale : locale.name;
  setDayjsLocale(localeName);

  const defaultLocales = { es, pt_BR: pt_br };

  if (typeof locale === 'string') return defaultLocales[locale];
  else {
    let customLocales = {};
    defaultsDeep(customLocales, { [localeName]: locale.translations }, defaultLocales);
    return customLocales[localeName];
  }
}
