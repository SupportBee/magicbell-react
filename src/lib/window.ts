/* eslint-disable no-restricted-globals */
import { useEffect, useRef, useState } from 'react';
import ResizeObserver from 'resize-observer-polyfill';

/**
 * Open a centered popup window.
 *
 * @param url Url of the floating window
 */
export function openWindow(url: string) {
  const top = (screen.height - 400) / 4;
  const left = (screen.width - 600) / 2;
  window.open(url, '', `width=600,height=400,scrollbars=no,top=${top},left=${left}`);
}

/**
 * Custom hook to observe the height of an HTML element.
 *
 * @param ref
 * @param initialHeight
 */
export function useHeight(ref, initialHeight) {
  const [height, setHeight] = useState(initialHeight);
  const resizeObserverRef = useRef() as any;

  useEffect(() => {
    resizeObserverRef.current = new ResizeObserver((entries: any = []) => {
      entries.forEach((entry) => {
        setHeight(entry.contentRect.height);
      });
    });

    if (ref.current) resizeObserverRef.current.observe(ref.current);

    return () => {
      if (resizeObserverRef.current) resizeObserverRef.current.disconnect();
    };
  }, [ref]);

  return height;
}
