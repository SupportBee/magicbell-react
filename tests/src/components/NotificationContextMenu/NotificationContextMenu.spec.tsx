import { useNotification } from '@magicbell/react-headless';
import { act, render, RenderResult } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import dayjs from 'dayjs';
import { Server } from 'miragejs';
import React from 'react';
import NotificationContextMenu from '../../../../src/components/NotificationContextMenu';
import { MagicBellThemeProvider } from '../../../../src/context/MagicBellThemeContext';
import { defaultTheme } from '../../../../src/context/Theme';
import { TranslationsProvider } from '../../../../src/context/TranslationsContext';
import { useLocale } from '../../../../src/lib/i18n';
import { sampleNotification } from '../../../factories/NotificationFactory';

describe('components', () => {
  describe('NotificationContextMenu', () => {
    let server;
    let view: RenderResult;

    beforeEach(() => {
      const { result } = renderHook(() => useNotification(sampleNotification));

      server = new Server({
        environment: 'test',
        urlPrefix: 'https://api.magicbell.com',
        timing: 50,
      });
      server.delete(`/notifications/${sampleNotification.id}`, '');
      server.post(`/notifications/${sampleNotification.id}/read`, '');
      server.post(`/notifications/${sampleNotification.id}/unread`, '');

      const translations = useLocale('pt_BR');

      view = render(
        <TranslationsProvider value={translations}>
          <MagicBellThemeProvider value={defaultTheme}>
            <NotificationContextMenu notification={result.current} />
          </MagicBellThemeProvider>
        </TranslationsProvider>,
      );
    });

    afterEach(() => {
      server.shutdown();
    });

    describe('render', () => {
      it('renders a menu with all items', () => {
        expect(view.container).toMatchSnapshot();
      });

      describe('default language', () => {
        it('renders a title in English', () => {
          const { result } = renderHook(() => useNotification(sampleNotification));

          view.rerender(
            <MagicBellThemeProvider value={defaultTheme}>
              <NotificationContextMenu notification={result.current} />
            </MagicBellThemeProvider>,
          );

          expect(view.container).toMatchSnapshot();
        });
      });
    });

    describe.skip('.toggleRead', () => {
      describe('the notification is read', () => {
        it('marks the notification as unread', async () => {
          const { waitForValueToChange, result } = renderHook(() =>
            useNotification({ ...sampleNotification, readAt: dayjs().unix() }),
          );
          view.rerender(
            <MagicBellThemeProvider value={defaultTheme}>
              <NotificationContextMenu notification={result.current} />
            </MagicBellThemeProvider>,
          );

          act(() => {
            const button = view.getByTestId('toggle-read');
            userEvent.click(button);
          });

          await waitForValueToChange(() => result.current.readAt);
        });
      });

      describe('the notification is unread', () => {
        it('marks the notification as read', async () => {
          const { waitForValueToChange, result } = renderHook(() =>
            useNotification({ ...sampleNotification, readAt: null }),
          );
          view.rerender(
            <MagicBellThemeProvider value={defaultTheme}>
              <NotificationContextMenu notification={result.current} />
            </MagicBellThemeProvider>,
          );

          act(() => {
            const button = view.getByTestId('toggle-read');
            userEvent.click(button);
          });

          await waitForValueToChange(() => result.current.readAt);
        });
      });
    });

    describe('.delete', () => {
      it.skip('deletes the notification', () => {
        const { result } = renderHook(() => useNotification(sampleNotification));
        const spy = jest.spyOn(result.current, 'delete');
        userEvent.click(view.getByTestId('delete'));

        expect(spy).toHaveBeenCalledTimes(1);
      });
    });
  });
});
