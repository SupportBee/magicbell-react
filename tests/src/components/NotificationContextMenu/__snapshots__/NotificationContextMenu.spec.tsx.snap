// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components NotificationContextMenu render default language renders a title in English 1`] = `
.emotion-0 {
  background: #FFFFFF!important;
  border-radius: 8px!important;
  box-shadow: 2px 2px 3px rgba(230, 230, 230, 0.55);
  color: #3A424D!important;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif!important;
  font-size: 14px!important;
  text-transform: none!important;
  min-width: 10em;
}

.emotion-0 a {
  display: block;
  padding: 0.75em 1.25em!important;
}

.emotion-0 a:hover {
  -webkit-text-decoration: underline!important;
  text-decoration: underline!important;
}

<div>
  <div
    class="emotion-0"
  >
    <a
      data-testid="toggle-read"
    >
      Mark as read
    </a>
    <a
      data-testid="delete"
    >
      Delete
    </a>
  </div>
</div>
`;

exports[`components NotificationContextMenu render renders a menu with all items 1`] = `
.emotion-0 {
  background: #FFFFFF!important;
  border-radius: 8px!important;
  box-shadow: 2px 2px 3px rgba(230, 230, 230, 0.55);
  color: #3A424D!important;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif!important;
  font-size: 14px!important;
  text-transform: none!important;
  min-width: 10em;
}

.emotion-0 a {
  display: block;
  padding: 0.75em 1.25em!important;
}

.emotion-0 a:hover {
  -webkit-text-decoration: underline!important;
  text-decoration: underline!important;
}

<div>
  <div
    class="emotion-0"
  >
    <a
      data-testid="toggle-read"
    >
      Marcar como lido
    </a>
    <a
      data-testid="delete"
    >
      Apagar
    </a>
  </div>
</div>
`;
